src/orange/detail/BIHMaker.cc

```diff
@@ -0,0 +1,329 @@
+
+        if (axes_centers[ax].size() < 2)
+        {
+            continue;
+        }
+        else
+        {
+            partition_axis = axis;
+            auto size = axes_centers[ax].size();
+
+            partition_location
+                = (axes_centers[ax][size / 2 - 1] + axes_centers[ax][size / 2])
+                  / 2;
+            break;
+        }
+    }
+    return {partition_axis, partition_location};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Divide bboxes into left and right branches based on a partition.
+ */
+BIHMaker::PairVecIndices BIHMaker::partition_bboxes(VecIndices const& indices,
+                                                    VecReal3 const& centers,
+                                                    Partition const& p)
+{
+    CELER_EXPECT(!indices.empty());
+    CELER_EXPECT(!centers.empty());
+    CELER_EXPECT(indices.size() == centers.size());
+
+    VecIndices left;
+    VecIndices right;
+
+    for (auto i : range(indices.size()))
+    {
+        if (centers[i][to_int(p.axis)] < p.location)
+        {
+            left.push_back(indices[i]);
+        }
+        else
+        {
+            right.push_back(indices[i]);
+        }
+    }
+
+    return std::make_pair(left, right);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Add leaf volume ids to a given node.
+ */
+void BIHMaker::make_leaf(BIHNode& node, VecIndices const& indices)
+{
+    CELER_EXPECT(!node);
+    CELER_EXPECT(!indices.empty());
+
+    node.vol_ids
+        = make_builder(storage_).insert_back(indices.begin(), indices.end());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the centers of each bounding box.
+ */
+BIHMaker::VecReal3 BIHMaker::centers(VecIndices const& indices)
+{
+    CELER_EXPECT(!indices.empty());
+    VecReal3 centers(indices.size());
+    for (auto i : range(indices.size()))
+    {
+        Real3 center;
+        auto bbox = bboxes_[indices[i].unchecked_get()];
+
+        for (auto axis : range(Axis::size_))
+        {
+            auto ax = to_int(axis);
+            center[ax] = (bbox.lower()[ax] + bbox.upper()[ax]) / 2;
+        }
+        centers[i] = center;
+    }
+    return centers;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create sorted and uniquified X, Y, Z values of bbox centers.
+ */
+BIHMaker::AxesCenters BIHMaker::axes_centers(VecReal3 const& centers)
+{
+    CELER_EXPECT(!centers.empty());
+
+    AxesCenters axes_centers{{}, {}, {}};
+
+    for (auto center : centers)
+    {
+        for (auto axis : range(Axis::size_))
+        {
+            auto ax = to_int(axis);
+            axes_centers[ax].push_back(center[ax]);
+        }
+    }
+
+    for (auto axis : range(Axis::size_))
+    {
+        auto ax = to_int(axis);
+        sort_and_uniquify(axes_centers[ax]);
```

Comment:
I don't think "uniquify" make sense algorithmically for a floating point–based algorithm. We'd still want to choose a partition point based on weighting by coincident centers, right?

Reply:
A)I was planning on adding a fairly large tolerance to uniquify (perhaps 1E-6) for reason #1 below, but I wanted to think about it more.
B) When writing all of this code, I found that no matter what choices you make in partitioning, you can always come up with a pathological counter example. I think it makes sense to uniquify for two reasons, with the first reason being the main one:

1) Let's say you have bbox centers at 0, 10, 10 10.001, 20, and 30. If you don't uniquify you are going to put a partition between 10 and 10.001. If your bounding boxes are 1x1x1 or something, this means that your lower and upper bounding interval overlap, meaning you will have to traverse both sides of the tree whenever you are in this region. It would be better to put the split at 15. Then the left bounding interval ends at 10.501 and the right bounding interval starts at 19.5. These intervals don't overlap, so you will always search one side or the other. The bottom line is BIH is not well-suited to partition bounding boxes with the same centers, so we should avoid trying.

2) By not uniquifying away coincident centers, you will get a more balanced tree, i.e. each partition will split the nodes roughly in half. If when sampling, points are equally likely to be in any bbox, a balanced tree is the best. However, this is not a good assumption for our application. I better assumption would be that all points in space are equally likely. This is kind of a murky argument, but I feel like if you don't uniquify, you end up "double counting" a region of space which may have a lot of overlapping bboxes in it, but is ultimately no more likely than any other similarly sized region in space to contain any given point.

Reply:
Maybe we need to do a little more research on the various partitioning algorithms before making a decision? I also suspect that the optimal hierarchy will be different for initialization versus ray tracing: the ray tracing heuristics I've seen tend to rely on a "surface area" based approach due to the probability of hitting it during ray tracing, but for point-in-volume initialization we'd want a volume based I think.

Reply:
There isn't much in the literature about partitioning schemes for BIH.  I have refactored things so that now I am testing every possible partition candidate in x, y, and z and choosing the best one with a cost function. Right now am using the standard surface area heuristic cost function (but this could be experimented with very easily). This cost function takes into account both the surface area of the children and the tree balance, i.e. how many bboxes are on either side of the partition. Note that there is still a `uniquify`, but now it is just serving to prevent testing the same partition candidate multiple times.